### PR TITLE
Fix Image Caching

### PR DESCRIPTION
**Pull request by kidroca**
## Description

This PR fixes some subtle inconsistencies around image caching and rendering the same image multiple times

Related to:
- upstream issue: https://github.com/necolas/react-native-web/issues/2492
- App issue: https://github.com/Expensify/App/issues/14278
- upstream PR: https://github.com/necolas/react-native-web/pull/2493

## Test Strategy

Use App's PR to test the changes (https://github.com/Expensify/App/pull/15663)
1. Log into an account on New Expensify
2. Navigate to a chat where you were the last person to send a message
3. Send a message in that chat
4. Confirm that the avatar on the left hand side of the message that you just sent does not flickers.

### Why No Unit Tests

Couldn't easily add unit tests for this change, because the main change is to the `ImageLoader.load` which doesn't have a test suit, and is mocked in the `Image` tests

Existing tests cover the changes in the `Image` component

### CODE REVIEW

File: packages/react-native-web/src/exports/Image/index.js [hunk 1]
```diff
@@ -211,24 +211,18 @@ const BaseImage: ImageComponent = React.forwardRef((props, ref) => {
...
-    if (uri != null) {
-      const isLoaded = ImageLoader.has(uri);
-      if (isLoaded) {
-        return LOADED;
-      }
-    }
-    return IDLE;
-  });
-
+  const [state, updateState] = React.useState(IDLE);
```
Comment (kidroca) [inline packages/react-native-web/src/exports/Image/index.js:214]: This is a state initializer, it runs only once, because of that the component does not behave correctly when we switch `source`, cases like 
- component is mounted with `null` source then changed to actual source that was already loaded - state was already initialized and is not updated to LOADED
- component changes source dynamically to some other source that was loaded - same problem as above

This causes flickers on some occasions

File: packages/react-native-web/src/exports/Image/index.js [hunk 2]
```diff
@@ -211,24 +211,18 @@ const BaseImage: ImageComponent = React.forwardRef((props, ref) => {
...
   const hiddenImageRef = React.useRef(null);
   const filterRef = React.useRef(_filterId++);
   const requestRef = React.useRef(null);
+  const uri = resolveAssetUri(source);
+  const isCached = uri != null && ImageLoader.has(uri);
   const shouldDisplaySource =
-    state === LOADED || (state === LOADING && defaultSource == null);
+    state === LOADED ||
+    isCached ||
+    (state === LOADING && defaultSource == null);
```
Comment (kidroca) [inline packages/react-native-web/src/exports/Image/index.js:225]: We can capture the old "state initializer" in a hook and run it on source change, so that if already loaded source is passed we can directly switch to LOADED state, but I find it simpler to always start from IDLE state and capture `isCached` like the proposed change. This way we've evaluated the correct result on the same render, while an use effect would only update state on the next render

File: packages/react-native-web/src/modules/ImageLoader/index.js [hunk 3]
```diff
@@ -74,12 +74,13 @@ let id = 0;
...
-  abort(requestId: number) {
-    let image = requests[`${requestId}`];
+  clear(requestId: number) {
+    const image = requests[`${requestId}`];
     if (image) {
       image.onerror = null;
       image.onload = null;
-      image = null;
+      ImageUriCache.remove(image.src);
+      image.src = '';
```
Comment (kidroca) [inline packages/react-native-web/src/modules/ImageLoader/index.js:83]: renamed this to `clear` to represent more closely that it's not just for aborting but actually a necessary cleanup, as it also updates the `ImageUriCache` now

Existing functionality is unaffected (besides the rename)

**Review by kidroca — COMMENTED** [review]
Added some notes to aid the reviewer

**Comment by mountiny** [comment]
Requesting @marcaaron and @Beamanator for review if you have time given the comment from Kidroca about previous PR in this library

**Comment by marcaaron** [comment]
Sorry, I've got to focus on some other things atm so unassigning myself as a reviewer.

**Comment by kidroca** [comment]
Bump @mountiny @Beamanator 
Some tickets are blocked by this PR 
- https://github.com/Expensify/App/issues/14278 (original issue)
- https://github.com/Expensify/App/issues/15922#issuecomment-1476650130 (would benefit from the change)

We're introducing a small change here so it should be easy to review
_(There are PR comments regarding the changes)_

File: packages/react-native-web/src/modules/ImageLoader/index.js [hunk 4]
```diff
@@ -74,12 +74,13 @@ let id = 0;
 const requests = {};
 
 const ImageLoader = {
-  abort(requestId: number) {
-    let image = requests[`${requestId}`];
+  clear(requestId: number) {
```
Comment (mountiny) [inline packages/react-native-web/src/modules/ImageLoader/index.js:77]: So I am wondering, we change this in our fork but is there any plan of changing it upstream? Will we have to maintain this diff from now on? If its just the naming that feels like unnecessary change if we wont get it merged upstream

**Comment by Beamanator** [comment]
Sorry @kidroca @mountiny I'm not working much this week so won't be able to review, but can try to get to this next week 🙃

**Review by mountiny — COMMENTED** [review]
@kidroca one quesion

File: packages/react-native-web/src/modules/ImageLoader/index.js (hunk 4 above)
Comment (mountiny) [reply packages/react-native-web/src/modules/ImageLoader/index.js:77]: Sorry haven't seen the linked PR before https://github.com/necolas/react-native-web/pull/2493

**Comment by mountiny** [comment]
@kidroca just to confirm, have you tested this in the App too? do you have a screen recording? We should test it for sure before going ahead

Oh I see its being pointed to the compiled code in this PR so I will go ahead and merge this

**Review by mountiny — APPROVED** [review]
Tests well in the linked App PR, lets do this https://github.com/Expensify/App/pull/15663

**Comment by kidroca** [comment]
Yep, i confirm the following 
- tested compiled code against App (videos available in App's PR)
- opened the same PR against the mainstream repo (there's also a linked issue with a generic bug reproduction sample)

Thanks for the review @mountiny 🫶
